src/dune_pkg/rev_store.ml

```diff
@@ -2,43 +2,97 @@ open Stdune
 open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
+module Scheduler = Dune_engine.Scheduler
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+type t =
+  { dir : Path.t
+  ; lock : Path.t
+  }
+
 type rev = Rev of string
 
-let equal { dir } t = Path.equal dir t.dir
+(* Async-inspired variation of [Fiber.repeat_until] *)
+let rec repeat_until_finished state f =
+  let* computation = f state in
+  match computation with
+  | `Repeat state -> repeat_until_finished state f
+  | `Finished result -> Fiber.return result
+;;
+
+let lock_limited flock lock timeout =
+  let timeout = max timeout 0. in
+  let sleep_duration = 0.1 in
+  repeat_until_finished timeout (fun timeout ->
+    if timeout > 0.
+    then (
+      match Flock.lock_non_block flock lock with
+      | Ok `Success as ok -> Fiber.return @@ `Finished ok
+      | Ok `Failure ->
+        let+ () = Scheduler.sleep sleep_duration in
+        `Repeat (timeout -. sleep_duration)
```

Comment:
One alternative is to keep a count of retries. That avoids float math weirdness like "can repeated substraction of 0.1 from 5. yield a negative number?". In that case the timeout is only used once so I think we could remove the parameter and hardcode 50 retries of 0.1 seconds. Just an idea

Reply:
Sure. I thought a time was a bit better of an API but it's unpleasantly continuous. This approach has the advantage that we can guarantee 50 iterations, but 50 times sleeping 0.1 doesn't mean that we stopped at 5s, it might've been more.

The main ugly thing is that this is still trying for some randomly picked period, but I can't think of an immediately better solution.